### PR TITLE
`General`: Add the possibility for instructors to define (un)enrollment periods

### LINKED ISSUE #334: Leave or hide courses
### Is your feature request related to a problem?
There is no way of leaving a course.

### Describe the solution you'd like
I would like to be able to leave a course as a student.

### PR DESCRIPTION
**Pull request by MaximilianAnzinger**
#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] This is a small issue that I tested locally and was confirmed by another developer on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/guidelines/development-process/#naming-conventions-for-github-pull-requests).
#### Server
- [x] **Important**: I implemented the changes with a very good performance and prevented too many (unnecessary) database calls.
- [x] I followed the [coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/).
- [x] I added multiple integration tests (Spring) related to the features (with a high test coverage).
- [x] I added pre-authorization annotations according to the [guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/#rest-endpoint-best-practices-for-authorization) and checked the course groups for all new REST Calls (security).
- [x] I documented the Java code using JavaDoc style.
#### Client
- [x] **Important**: I implemented the changes with a very good performance, prevented too many (unnecessary) REST calls and made sure the UI is responsive, even with large data.
- [x] I followed the [coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client/).
- [x] Following the [theming guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-design/), I specified colors only in the theming variable files and checked that the changes look consistent in both the light and the dark theme.
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-tests/).
- [x] I documented the TypeScript code using JSDoc style.
- [x] I added multiple screenshots/screencasts of my UI changes.
- [x] I translated all newly inserted strings into English and German.

### Motivation and Context
If students decide against participating in a course, there is currently no way to disenroll from a course.
Resolves #334 and #6654.

### Description
This PR adds
- the possibility for instructors to configure a time period for students to enroll in the course
- the possibility for instructors to allow students to unenroll from a course within a specified time period

Applies new naming convention for `(un)enrollment`. See #6565 .

### Steps for Testing

Prerequisites:
- 1 Admin
- 1 Instructor
- 1 Students
- 1 Programming Exercise (allow online editor to simplify the process)

1. Create a course and make sure you can configure (un)enrollment and you get adequate error messages if the selected time period is not valid (e.g. enrollment still possible after unenrollment end date or unenrollment possible after course end date)
2. Edit the course and make sure you still can configure the (un)enrollment dates as expected.
3. Create a programming exercise
4. Test that a student can not enroll in online courses, before enrollment starts, and after enrollment ends (adjust the course accordingly using the instructor OR create multiple courses)
5. Test that a student can enroll during the enrollment period
6. Make sure the student can see the correct course information (course start & end dates, enrollment start & end dates, unenrollment end date) within the course (`https://artemis.ase.in.tum.de/courses/<course-id>/exercises`)
7. Participate in the programming exercise and submit some changes
8. Try to unenroll from the course and make sure the modal displays correctly if the student can enroll again (enrollment period still in progress) or if it is not possible to reenroll afterward (past enrollment end date).
9. Finish the unenrollment process and make sure the course is no longer visible to you
10. Make sure the instructor can still see the submission(s) of the student
11. Enroll the student again and make sure that the submission(s) still exist

#### Performance Review
- [x] I confirm that the client changes (in particular related to REST calls and UI responsiveness) are implemented with a very good performance 
- [x] I confirm that the server changes (in particular related to database calls) are implemented with a very good performance
#### Code Review
- [x] Code Review 1
- [x] Code Review 2
#### Manual Tests
- [x] Test 1
- [ ] Test 2

### Test Coverage

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|-----------------------------:|
| Constants.java | 100% | ✅                           |
| Course.java | 89% | ✅              |
| CourseRepository.java | 81% | ✅              |
| AuthorizationCheckService.java | 91% | ✅              |
| CourseService.java | 93% | ✅              |
| CourseResource.java | 95% | ✅              |
| AdminCourseResource.java | 100% | ✅              |
| course-management.service.ts | 87% | ✅              |
| course-update.component.ts | 81% | ✅              |
| course.model.ts | 100% | ✅              |
| course-detail.component.ts | 92% | ✅              |
| course-exercises.component.ts | 99% | ✅              |
| course-unenrollment-modal.component.ts | 96% | ✅              |

### Screenshots

Course_Management_Desktop

Course_Info

Screenshot 2023-05-27 at 13 44 03

### CODE REVIEW

File: src/main/java/de/tum/in/www1/artemis/domain/Course.java
Comment (b-fein) [inline src/main/java/de/tum/in/www1/artemis/domain/Course.java:172]:
```suggestion
    private boolean disenrollmentEnabled = false;
```
Early suggestion: Please make this a non-nullable data type and also set a default value in the database definition.

Booleans with `null` are tricky since there are multiple ways to handle that as `true` or `false` depending on context, and all comparisons have to be made using `Booleans.TRUE.equals(…)` or similar.

Comment (MaximilianAnzinger) [reply src/main/java/de/tum/in/www1/artemis/domain/Course.java]: Thx for the suggestion. I changed it immediately: [commit](https://github.com/ls1intum/Artemis/pull/6577/commits/4d5af9382f28f2bb615850321a7b1b6b72b38d96)

File: src/main/resources/config/liquibase/changelog/20230514121500_changelog.xml
Comment (b-fein) [inline src/main/resources/config/liquibase/changelog/20230514121500_changelog.xml:16]:
```suggestion
            <column name="disenrollment_enabled" type="boolean"  defaultValueBoolean="false">
```

File: src/main/resources/config/liquibase/master.xml
Comment (JohannesStoehr) [inline src/main/resources/config/liquibase/master.xml:33]: Your new file should be at the end of the change list

File: src/main/webapp/app/overview/course-exercises/course-exercises.component.ts
Comment (JohannesStoehr) [inline src/main/webapp/app/overview/course-exercises/course-exercises.component.ts:239]:
```suggestion
        return !!this.course?.unenrollmentEnabled && dayjs().isBefore(this.course.unenrollmentEndDate);
```
Or
```suggestion
        return this.course?.unenrollmentEnabled && dayjs().isBefore(this.course.unenrollmentEndDate) ?? false;
```

File: src/main/webapp/i18n/de/course.json
Comment (JohannesStoehr) [inline src/main/webapp/i18n/de/course.json:103]: "Ausschreibung" is a bit confusing vor me, since it could also be something like an advertisement.
But I cannot find any better translation, maybe someone else can think of one. Otherwise leave it as is

**Review by JohannesStoehr — CHANGES_REQUESTED** [review]
Left some comments, looks good overall! 👍 
I'd suggest to ask the student to type in the name of the course when unenrolling like when deleting exercises for example. This could save students and instructors some headaches...

File: src/main/webapp/i18n/de/course.json
Comment (MaximilianAnzinger) [reply src/main/webapp/i18n/de/course.json:103]: I agree but the only better alternative I can think of would be "Abmeldung" which conflicts with the new convention.

Comment (b-fein) [reply src/main/webapp/i18n/de/course.json:103]: The Stud.IP LMS (alternative to Moodle) uses ‘eintragen’ and ‘austragen’ for course (un-)enrolment. That might be a good alternative.

**Review by JohannesStoehr — DISMISSED** [review]
Thanks for addressing my feedback, code looks good to me

**Review by SabrinaGlatz — DISMISSED** [review]
tested manually on ts1. It works as expected.

**Review by rriyaldhi — DISMISSED** [review]
Tested on TS11 and works as described.

**Review by laadvo — DISMISSED** [review]
Tested on ts11 during the testing session, the enrollement and unenrollment dates work as expected

**Review by frederik-kt — APPROVED** [review]
Tested on TS11. Worked as expected. It would be nice to also have the enrollment/unenrollment dates in the overview page for the instructor.

**Review by greekgeek1 — DISMISSED** [review]
Tested on ts3, works as expected.

**Review by JohannesStoehr — DISMISSED** [review]
Code of latest commits after testing session look good to me

File: src/main/webapp/i18n/de/course.json
Comment (MaximilianAnzinger) [reply src/main/webapp/i18n/de/course.json]: I changed it to `austragen` as @b-fein suggested.

**Review by JohannesStoehr — DISMISSED** [review]
Approve code after latest commits

**Review by SabrinaGlatz — APPROVED** [review]
Tested on ts1, works as expected

**Review by tobias-lippert — DISMISSED** [review]
Tested in testing session on ts2. Everything worked as expected. Only one error message was a bit misleading but this is non-blocking and will be addressed in a follow-up.

**Review by rriyaldhi — APPROVED** [review]
Reapproved.

**Review by pellumbbaboci — APPROVED** [review]
tested on ts2 during testing session, works as expected 👍

File: src/main/webapp/app/overview/course-exercises/course-exercises.component.html
Comment (rriyaldhi) [inline src/main/webapp/app/overview/course-exercises/course-exercises.component.html:197]:
```suggestion
                <div class="col text-end">
```

**Review by rriyaldhi — APPROVED** [review]
Code looks good. Just left small optional code suggestions.

**Review by rriyaldhi — DISMISSED** [review]
Reapprove code and functionality. The latest changes do not impact the functionality.

**Review by JohannesStoehr — DISMISSED** [review]
Reapprove code. Changes from other code review look reasonable

**Review by SabrinaGlatz — APPROVED** [review]

**Review by asliayk — APPROVED** [review]

**Review by asliayk — DISMISSED** [review]
tested on ts2, worked well

**Review by DemjenDavid — APPROVED** [review]
Tested on ts3, all functionalities worked as expected

**Review by rriyaldhi — APPROVED** [review]
Reapprove after test fix

**Review by pellumbbaboci — APPROVED** [review]
Re-approve after fix

**Review by asliayk — APPROVED** [review]
reapproved

**Review by SabrinaGlatz — APPROVED** [review]

**Review by JohannesStoehr — APPROVED** [review]
Reapprove code after test fix

**Review by tobias-lippert — APPROVED** [review]
Reapprove

**Review by DemjenDavid — DISMISSED** [review]
Reapprove, tested on ts3

**Review by greekgeek1 — DISMISSED** [review]
Reapprove

**Review by JohannesStoehr — APPROVED** [review]
Mergecommits look good to me

**Review by rriyaldhi — APPROVED** [review]
Reapprove after merge conflict

**Review by tobias-lippert — APPROVED** [review]
Retested on ts2. Successfully configured enrollment and unenrollment periods and was able to enroll as a student and to unenroll.

**Review by pellumbbaboci — APPROVED** [review]
re-tested on ts2, works as expected. re-approve

**Review by greekgeek1 — APPROVED** [review]
Reapprove.

**Pull request merged**
